### PR TITLE
Developing a JSF Portlet Application

### PR DESCRIPTION
**Pull request by codyhoag**
Hey Kyle,

Can you review the *Developing a JSF Portlet Application* for accuracy? Note, this generalized article is targeted for JSF developers who are migrating their web app to work on Liferay Portal. Let me know if there are any additional things we should point out. Thanks!

### CODE REVIEW

File: developer/appdev/articles/01-web-front-ends/06-jsf-portlet/01-jsf-portlet-intro.markdown [hunk 1]
```diff
@@ -6,4 +6,87 @@ header-id: jsf-portlet
...
+- JSF is a unique framework in that it provides a UI component model that makes
+  it easy to develop sophisticated, rich user interfaces.
+- JSF has built-in Ajax functionality that provides automatic updates to the
+  browser by replacing elements in the DOM.
+- JSF is designed with many extension points that make a variety of integrations
+  possible.
+- There are several JSF component suites available including Liferay Faces
+  Alloy, Primefaces, ICEfaces, and RichFaces. Each of these component suites
+  fortify JSF with a variety of UI components and complimentary technologies
+  such as Ajax Push.
```
Comment (stiemannkj1) [inline developer/appdev/articles/01-web-front-ends/06-jsf-portlet/01-jsf-portlet-intro.markdown]: Remove the part about Ajax push as it's unlikely to work in Liferay 7.0 and it's not really that popular anymore I think.

File: developer/appdev/articles/01-web-front-ends/06-jsf-portlet/01-jsf-portlet-intro.markdown [hunk 2]
```diff
@@ -6,4 +6,87 @@ header-id: jsf-portlet
...
+- Since JSF is a stateful technology, the framework encapsulates the
+  complexities of managing application state so the developer doesn't have
+  to write state management code. It is also possible to use JSF in a stateless
+  manner, but some of the features of application state management become
+  effectively disabled.
+
+There are some reasons not to use JSF. For example, if you are a front-end
+developer who makes heavy use of HTML/CSS/JavaScript, you might find that JSF UI
+components render HTML in a manner that gives you less control over the overall
+HTML document. Sticking with JavaScript and leveraging AlloyUI or some other
```
Comment (stiemannkj1) [inline developer/appdev/articles/01-web-front-ends/06-jsf-portlet/01-jsf-portlet-intro.markdown]: AlloyUI is deprecated as far as I know and should not be promoted. Check with the front-end folks to make sure.

File: developer/appdev/articles/01-web-front-ends/06-jsf-portlet/02-developing-a-jsf-portlet-application.markdown [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+5.  Add your frontend files (e.g., CSS, JS, etc.) that shouldn't be accessed
+    directly by the browser to the `webapp/WEB-INF/resources/` folder. For
+    example, your web app's CSS files would reside in the
+    `webapp/WEB-INF/resources/css` folder.
+
+6.  Add your image files to the `webapp/resources/images` folder.
+
+7.  Add localized messages to the `resources/i18n.properties` file. The messages
+    in the `i18n.properties` file can be accessed via the Expression Language
+    using the implicit `i18n` object provided by the Liferay Faces Util class.
```
Comment (stiemannkj1) [inline developer/appdev/articles/01-web-front-ends/06-jsf-portlet/02-developing-a-jsf-portlet-application.markdown]: ~the Liferay Faces Util class~ -> Liferay Faces Util

File: developer/appdev/articles/01-web-front-ends/06-jsf-portlet/02-developing-a-jsf-portlet-application.markdown [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+    It describes the portlet application's packaging and version information
+    and specifies any required OSGi metadata. For example, this
+    `liferay-plugin-package.properties` snippet tells the OSGi container not to
+    scan for CDI annotations.
+
+    ```properties
+    -cdiannotations:
+    ```
+
+    This is useful if you want to include your own CDI implementation with your
```
Comment (stiemannkj1) [inline developer/appdev/articles/01-web-front-ends/06-jsf-portlet/02-developing-a-jsf-portlet-application.markdown]: I like the text here, but it's technically required to use `-cdiannotations:` with JSF + CDI in Liferay. Devs won't be able to take advantage of Liferay's included CDI impl until we add thin-war support. So perhaps this should be reworded to reflect that it's a requirement for JSF + CDI portlets (not necessary for non-CDI, JSF portlets though).

**Review by stiemannkj1 — CHANGES_REQUESTED** [review]
@codyhoag, the article looks fantastic. It's a really brilliant idea that I think will help lots of devs and it's really clear and well written and technically accurate (especially the benefits/drawbacks of JSF section). I noticed 4 minor issues, but once those are fixed, I think this is ready to be published.

**Comment by codyhoag** [comment]
Thanks for the quick feedback, @stiemannkj1! I've incorporated your feedback and sent on for publication.

**Pull request closed without merging**